### PR TITLE
Cert pinning on the oautn2 client.

### PR DESCRIPTION
**Pull request by chogorma**

### CODE REVIEW

File: Release/include/cpprest/oauth2.h [hunk 1]
```diff
@@ -66,6 +66,8 @@ class oauth2_strings
 namespace experimental
 {
 
+using CertificateChainCallBackFunction = std::function<bool(const utility::string_t&, const std::vector<std::vector<unsigned char>>&)>;
```
Comment (deeringc) [inline Release/include/cpprest/oauth2.h:69]: Cant we just refer to the type that's used already by the http client code?

File: Release/include/cpprest/oauth2.h [hunk 2]
```diff
@@ -479,6 +481,17 @@ class oauth2_config
...
+    /// <summary>
+    /// Set the certificate chain callback to be used by the http client.
+    /// </summary>
+    void set_user_certificate_chain_callback(const CertificateChainCallBackFunction& callback) { m_certificate_chain_callback = callback; }
+
+    /// <summary>
+    /// Get the cert chain callback.
+    /// </summary>
+    /// <returns>A reference to cert chain callback user by the client.</returns>
+    const CertificateChainCallBackFunction& user_certificate_chain_callback() { return m_certificate_chain_callback; }
```
Comment (deeringc) [inline Release/include/cpprest/oauth2.h:493]: Not sure about returning const ref to this, what does it achieve in this case for std:function?

File: Release/include/cpprest/oauth2.h (hunk 1 above)
Comment (chogorma) [reply Release/include/cpprest/oauth2.h:69]: We don't want to include http_client.h in this file.

File: Release/include/cpprest/oauth2.h (hunk 2 above)
Comment (chogorma) [reply Release/include/cpprest/oauth2.h:493]: Don't want anything to change the callback.

File: Release/src/http/client/x509_cert_utilities_win32.cpp [hunk 3]
```diff
@@ -67,23 +65,40 @@ bool verify_X509_cert_chain(const std::vector<std::string> &certChain, const std
...
     chain_context chain;
     if (!CertGetCertificateChain(
         nullptr,
         cert.get(),
         nullptr,
-        nullptr,
+        caMemStore,
         &params,
         CERT_CHAIN_REVOCATION_CHECK_CHAIN,
         nullptr,
```
Comment (coliemcgarry) [inline Release/src/http/client/x509_cert_utilities_win32.cpp:104]: Need to 'free' the store in the 'return false'

**Review by coliemcgarry — APPROVED** [review]
Just fix the release of the Certificate Store

**Pull request merged**